api/src/routes/ical.ts

```diff
@@ -0,0 +1,73 @@
+import { ICalCalendar } from "ical-generator";
+import { Request, Response, Handler } from "express";
+import { Pool } from "pg";
+import slugify from "slugify";
+
+type CtfRow = {
+  id: number;
+  title: string;
+  start_time: string;
+  end_time: string;
+  ctf_url: string;
+  description: string;
+};
+
+type IcalPasswordRow = {
+  ical_password: string;
+};
+
+export function icalRoute(pool: Pool): Handler {
+  async function checkIcalPassword(
+    userPass: string | undefined
+  ): Promise<boolean> {
+    const r = await pool.query<IcalPasswordRow>(
+      "SELECT ical_password FROM ctfnote.settings"
+    );
+    const db_password = r.rows[0].ical_password;
+    // If the password is null or empty allow any user
+    if (!db_password) return true;
+    return db_password === userPass;
+  }
+
+  async function getCtfs(): Promise<CtfRow[]> {
+    const r = await pool.query<CtfRow>(
+      "SELECT id, title, start_time, end_time, ctf_url, description FROM ctfnote.ctf"
+    );
+
+    return r.rows;
+  }
+
+  return async function (req: Request, res: Response): Promise<void> {
+    const { key } = req.query;
+
+    if (
+      !(typeof key == "string" || key == undefined) ||
+      !(await checkIcalPassword(key))
+    ) {
+      res.status(403);
+      res.send("Forbidden\n");
+      return;
+    }
+
+    const cal = new ICalCalendar();
+    const ctfs = await getCtfs();
+
+    for (const ctf of ctfs) {
+      // Assumes that CTFNote is hosted at the root of the domain
+      const ctf_url = new URL(
+        `/#/ctf/${ctf.id}-${slugify(ctf.title)}/info`,
```

Comment:
See comment about the `slugify` package.
Can we retrieve `ctf.slug` instead ? See `/front/src/components/Utils/CtfNoteLink.vue`, line 29.

Reply:
I think it is perfectly to regenerate it in the backend since it is nice to see which CTF you are going to visit when clicking the link. Do you agree @XeR?

Reply:
So how should I approach / fix / resolve this?

Reply:
I think you can leave it like this or if it is purely visual then you may also remove it.

Reply:
@frereit, can you confirm that the slugify in the backend is purely visual? So that means that if we remove the slugify part from the URL, the link still works? If so, I think it is preferred to remove the dependency and therefore the slugified part of the URL. So we only show a URL with an ID in it. I don't think this is a problem since the event will already contain the name of the event.

Reply:
Yes, the slug is purely visual, however just the id is not enough, so a valid URL for CTF with id 20 needs to match`/\/#\/ctf\/20-.+\//`. Just `/#/ctf/20/` wouldn't work. I haven't looked into this, but should be trivial to change the logic to allow id-only links. If you want, I can do that.

Reply:
I think that is the best solution. If you could change it to accept only IDs, that would be very nice!